modeshape-jcr/src/main/java/org/modeshape/jcr/locking/StandaloneLockingService.java

```diff
@@ -0,0 +1,123 @@
+        CheckArg.isNonNegative(lockTimeoutMillis, "lockTimeoutMillis");
+        this.lockTimeoutMillis = lockTimeoutMillis;
+    }
+
+    @Override
+    public boolean tryLock(long time, TimeUnit unit, String... names) {
+        Set<ReentrantLock> successfullyLocked = new HashSet<>();
+        for (String name : names) {
+            LOGGER.debug("Attempting to lock {0}", name);
+            ReentrantLock lock = null;
+            boolean success = false;
+            try {
+                lock = locksByName.computeIfAbsent(name, (lockName) -> new ReentrantLock());
+                success = time > 0 ? lock.tryLock(time, unit) : lock.tryLock();
+            } catch (InterruptedException e) {
+                LOGGER.debug("Thread {0} received interrupt request while waiting to acquire lock '{1}'", 
+                             Thread.currentThread().getName(), name);
+                Thread.currentThread().interrupt();
+            } catch (Exception e) {
+                LOGGER.debug(e, "Unexpected exception while attempting to lock '{0}'", name);
+            }
+            // if lock acquisition was not successful (for whatever reason) revert the lock
+            if (!success) {
+                LOGGER.debug("Unable to acquire lock on {0}. Reverting back the already obtained locks: {1}", name,
+                             successfullyLocked);
+                    successfullyLocked.stream().forEach(ReentrantLock::unlock);
+                return false;
+            }
+            LOGGER.debug("{0} locked successfully", name);
+            successfullyLocked.add(lock);
+        }
+        return true;
+    }
+
+    @Override
+    public List<String> unlock(String... names) {
+        return Arrays.stream(names)
+                     .map(this::unlock)
+                     .filter(Optional::isPresent)
+                     .map(Optional::get)
+                     .collect(Collectors.toList());
+    }
+    
+    private Optional<String> unlock(String name) {
+        LOGGER.debug("Attempting to unlock {0}", name);
+        ReentrantLock lock = locksByName.get(name);
+        if (lock == null) {
+            return Optional.of(name);
+        }
+        try {
+            lock.unlock();
+        } catch (IllegalMonitorStateException e) {
+            LOGGER.warn(JcrI18n.warnAnotherThreadAttemptingToUnlock, Thread.currentThread().getName(), lock);
+        }
+        if (lock.isLocked()) {
+            LOGGER.debug("Lock '{0}' is still locked with hold count '{1}'", name, lock.getHoldCount());
+            return Optional.of(name);
+        } else {
+            LOGGER.debug("'{0}' unlocked", name);
+            // if there is no contention for this lock, remove it
+            // otherwise leave it because the last queued thread will remove it once it completes
+            if (!lock.hasQueuedThreads()) {
+                locksByName.remove(name);
+            }
+            return Optional.empty();
```

Comment:
I agree; the above are good points with regards to the fact that the logic from `unlock` should be performed atomically (hence the `computeIfPresent` method).